src/shell/xdg/popup.rs

```diff
@@ -80,46 +78,45 @@ impl Popup {
         position: &xdg_positioner::XdgPositioner,
         qh: &QueueHandle<D>,
         surface: impl Into<Surface>,
-        wm_base: &impl ProvidesBoundGlobal<XdgWmBase, 4>,
+        wm_base: &impl ProvidesBoundGlobal<xdg_wm_base::XdgWmBase, 4>,
     ) -> Result<Popup, GlobalError>
     where
         D: Dispatch<xdg_surface::XdgSurface, PopupData>
             + Dispatch<xdg_popup::XdgPopup, PopupData>
             + 'static,
     {
-        // We really need an Arc::try_new_cyclic function to handle errors during creation.
-        // Emulate that function by creating an Arc containing None in the error case, and use the
-        // closure's context to pass the real error back to the caller so the invalid Arc is never
-        // returned.
-        let mut err = Ok(());
+        let surface = surface.into();
+        let wm_base = wm_base.bound_global()?;
+        let freeze = qh.freeze();
         let inner = Arc::new_cyclic(|weak| {
-            let surface =
-                XdgShellSurface::new(wm_base, qh, surface, PopupData { inner: weak.clone() })
-                    .map_err(|e| err = Err(e))
-                    .ok()?;
-            let xdg_popup = surface
-                .xdg_surface()
-                .get_popup(parent, position, qh, PopupData { inner: weak.clone() })
-                .map_err(|e| err = Err(e.into()))
-                .ok()?;
+            let xdg_surface = wm_base.get_xdg_surface(
+                surface.wl_surface(),
+                qh,
+                PopupData { inner: weak.clone() },
+            );
+            let surface = XdgShellSurface { surface, xdg_surface };
+            let xdg_popup = surface.xdg_surface().get_popup(
+                parent,
+                position,
+                qh,
+                PopupData { inner: weak.clone() },
+            );
 
-            Some(PopupInner {
+            PopupInner {
                 surface,
                 xdg_popup,
                 pending_position: (AtomicI32::new(0), AtomicI32::new(0)),
                 pending_dimensions: (AtomicI32::new(-1), AtomicI32::new(-1)),
                 pending_token: AtomicU32::new(0),
                 configure_state: AtomicU32::new(PopupConfigure::STATE_NEW),
-            })
+            }
         });
-        // This assert checks that err was set properly above; it should be impossible to trigger,
-        // so it's not a run-time assert.
-        debug_assert!(inner.is_some() || err.is_err());
-        err.map(|()| Popup { inner })
+        drop(freeze);
+        Ok(Popup { inner })
     }
 
     fn inner(&self) -> &PopupInner {
```

Comment:
This function is effectively redundant now.

Reply:
Which function? `from_surface` is not redundant with `new`.

Reply:
The `inner` functions

Reply:
Ah. The github quote function makes this very unclear: it shows lines 117-121, and your comment is about line 121-123.